presto-main/src/main/java/com/facebook/presto/operator/scalar/KleinbergBurstDetectionFunction.java

```diff
@@ -0,0 +1,212 @@
+     */
+    private Block kleinberg(Block timeseries, double s, double gamma)
+    {
+        // The number of events [N].
+        int numEntries = timeseries.getPositionCount();
+
+        // Calculate time intervals between successive events.
+        double[] intervals = new double[numEntries - 1];
+        // The minimum interval.
+        double delta = Double.MAX_VALUE;
+        for (int i = 0; i < numEntries - 1; ++i) {
+            double timeCurrent = DoubleType.DOUBLE.getDouble(timeseries, i);
+            double timeNext = DoubleType.DOUBLE.getDouble(timeseries, i + 1);
+            // 'timeseries' should be sorted in ascending order.
+            if (timeCurrent > timeNext) {
+                return null;
+            }
+            intervals[i] = timeNext - timeCurrent;
+            delta = Math.min(delta, intervals[i]);
+        }
+
+        // The time length of the whole timeseries [T].
+        double timeLength = DoubleType.DOUBLE.getDouble(timeseries, numEntries - 1) - DoubleType.DOUBLE.getDouble(timeseries, 0);
+        if (timeLength <= 0) {
+            return null;
+        }
+
+        // The upper limit of burst levels [K].
+        int burstLevelUpperLimit = (int) (Math.ceil(1 + Math.log(timeLength / delta) / Math.log(s)));
+        if (burstLevelUpperLimit == 0) {
+            return null;
+        }
+
+        double[] alpha = new double[burstLevelUpperLimit];  // Event rate at each burst level.
+        double[] alphaLn = new double[burstLevelUpperLimit];
+
+        alpha[0] = numEntries / timeLength;  // Average event rate.
+        alphaLn[0] = Math.log(alpha[0]);
+
+        for (int i = 1; i < burstLevelUpperLimit; ++i) {
+            alpha[i] = s * alpha[i - 1];
+            alphaLn[i] = Math.log(alpha[i]);
+        }
+
+        // Initialize.
+        Node[] q = new Node[burstLevelUpperLimit];
+        double[] costs = new double[burstLevelUpperLimit];
+        Arrays.fill(costs, Double.MAX_VALUE);
+        costs[0] = 0;
+
+        // Start optimization.
+        double gammalogn = gamma * Math.log((double) numEntries);
+        for (double interval : intervals) {
+            Node[] qNew = new Node[burstLevelUpperLimit];
+            double[] costsNew = new double[burstLevelUpperLimit];
+            for (int i = 0; i < burstLevelUpperLimit; ++i) {
+                double[] c = new double[burstLevelUpperLimit];
+                for (int j = 0; j < burstLevelUpperLimit; ++j) {
+                    c[j] = costs[j] + tau(gammalogn, j, i);
+                }
+                int jMin = findSmallestIndex(c);
+                // Store the cost for setting the burst level i.
+                costsNew[i] = -logf(alpha, alphaLn, i, interval) + c[jMin];
+                qNew[i] = new Node(i, q[jMin]);
+            }
+            q = qNew;
+            costs = costsNew;
+        }
+
+        // Build output structure.
+        List<Integer> bursts = new ArrayList<Integer>(Collections.nCopies(numEntries, 0));
```

Comment:
The type parameter for `new ArrayList<>(...)` is probably not needed. Also, static import `nCopies`.